src/Servers/HttpSys/test/FunctionalTests/ResponseBodyTests.cs

```diff
@@ -368,6 +368,43 @@ public async Task ResponseBody_WriteAsync_TriggersOnStarting()
         }
     }
 
+    [ConditionalTheory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public async Task ResponseBody_ZeroLengthTrailingWrite_Success(bool setContentLength)
+    {
+        string address, fault = null;
+        using (Utilities.CreateHttpServer(out address, async httpContext =>
+        {
+            byte[] data = Encoding.UTF8.GetBytes("hello, world");
+            if (setContentLength)
+            {
+                httpContext.Response.ContentLength = data.Length;
+            }
+            var body = httpContext.Response.Body;
+            await body.WriteAsync(data);
+            try
+            {
+                await body.FlushAsync();
+                await body.WriteAsync(Array.Empty<byte>());
+            }
+            catch (Exception ex)
+            {
+                // in content-length scenarios, server-side faults after
+                // the payload would not be observed
+                Volatile.Write(ref fault, ex.Message);
+            }
+        }))
+        {
+            var response = await SendRequestAsync(address);
+            var payload = await response.Content.ReadAsByteArrayAsync();
+            Assert.Equal("hello, world", Encoding.UTF8.GetString(payload));
+        }
+
+        await Task.Delay(50);
+        Assert.Null(fault);
```

Comment:
Instead of Task.Delay, what about setting a TCS?

```
            try
            {
                await body.FlushAsync();
                await body.WriteAsync(Array.Empty<byte>());
            }
            catch (Exception ex)
            {
                // in content-length scenarios, server-side faults after
                // the payload would not be observed
                tcs.SetException(ex);
                return;
            }
            tcs.SetResult();
```

Then await the tcs at the end of the test.

Reply:
Cleaner and more reliable. Will do.